feat: Added new MCP server for Twilio integration

raghavendra2222 opened the pull request:
## Description

This pull request submits the take-home coding assignment. It introduces a new, fully functional Model Context Protocol (MCP) server for the Twilio API.

The server provides an atomic tool `send_sms` that allows an AI agent to send SMS messages. All required deliverables are included in the `integrations/twilio/` directory.

## Related issue

N/A

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New MCP feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please specify)

## How has this been tested?

The server has been tested manually by running it locally and calling the `/tools/send_sms` endpoint using a `curl` command.

The successful execution, including the server logs and a confirmation of a received SMS, is documented in the "Proof of Correctness" screenshot located in the `integrations/twilio/` directory.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (manual test via curl)
- [x] New and existing tests pass locally with my changes

zihaolin96 committed:
fix google drive

zihaolin96 committed:
fix salesforce

zihaolin96 committed:
fix google docs

zihaolin96 committed:
fix google sheets

zihaolin96 committed:
Merge pull request #75 from Klavis-AI/fix/0610

Fix/mcp issue

zihaolin96 committed:
add confluence

zihaolin96 committed:
add docker

zihaolin96 committed:
update

zihaolin96 committed:
Merge pull request #76 from Klavis-AI/feat/confluence_update

feat/update confluence

dependabot[bot] committed:
Bump brace-expansion from 1.1.11 to 1.1.12 in /mcp_servers/firecrawl

Bumps [brace-expansion](https://github.com/juliangruber/brace-expansion) from 1.1.11 to 1.1.12.
- [Release notes](https://github.com/juliangruber/brace-expansion/releases)
- [Commits](https://github.com/juliangruber/brace-expansion/compare/1.1.11...v1.1.12)

---
updated-dependencies:
- dependency-name: brace-expansion
  dependency-version: 1.1.12
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

zihaolin96 committed:
Merge pull request #77 from Klavis-AI/dependabot/npm_and_yarn/mcp_servers/firecrawl/brace-expansion-1.1.12

Bump brace-expansion from 1.1.11 to 1.1.12 in /mcp_servers/firecrawl

zihaolin96 committed:
add create space

zihaolin96 committed:
Merge pull request #78 from Klavis-AI/fix/attio_confluence

fix/attio and confluence

dependabot[bot] committed:
Bump brace-expansion from 1.1.11 to 1.1.12 in /mcp_servers/attio

Bumps [brace-expansion](https://github.com/juliangruber/brace-expansion) from 1.1.11 to 1.1.12.
- [Release notes](https://github.com/juliangruber/brace-expansion/releases)
- [Commits](https://github.com/juliangruber/brace-expansion/compare/1.1.11...v1.1.12)

---
updated-dependencies:
- dependency-name: brace-expansion
  dependency-version: 1.1.12
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

zihaolin96 committed:
Merge pull request #79 from Klavis-AI/dependabot/npm_and_yarn/mcp_servers/attio/brace-expansion-1.1.12

Bump brace-expansion from 1.1.11 to 1.1.12 in /mcp_servers/attio

dependabot[bot] committed:
Bump brace-expansion from 2.0.1 to 2.0.2 in /mcp_servers/resend

Bumps [brace-expansion](https://github.com/juliangruber/brace-expansion) from 2.0.1 to 2.0.2.
- [Release notes](https://github.com/juliangruber/brace-expansion/releases)
- [Commits](https://github.com/juliangruber/brace-expansion/compare/v2.0.1...v2.0.2)

---
updated-dependencies:
- dependency-name: brace-expansion
  dependency-version: 2.0.2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

zihaolin96 committed:
add gong mcp server

zihaolin96 committed:
Merge pull request #80 from Klavis-AI/dependabot/npm_and_yarn/mcp_servers/resend/brace-expansion-2.0.2

Bump brace-expansion from 2.0.1 to 2.0.2 in /mcp_servers/resend

xiangkaiz committed:
fix and add more features.

xiangkaiz committed:
Fix bugs.

xiangkaiz committed:
Fix bugs.

xiangkaiz committed:
Merge pull request #82 from Klavis-AI/attio

fix and add more features for attio

zihaolin96 committed:
add clickup

zihaolin96 committed:
Merge pull request #83 from Klavis-AI/feat/clickup

add clickup

zihaolin96 committed:
add motion

zihaolin96 committed:
Merge pull request #84 from Klavis-AI/feat/motion

feat/add motion

zihaolin96 committed:
Merge pull request #81 from Klavis-AI/feature/add_gong

feat/add gong mcp server

zihaolin96 committed:
add together and fireworks

zihaolin96 committed:
llamaindex

zihaolin96 committed:
update

zihaolin96 committed:
Merge pull request #86 from Klavis-AI/feat/integration_example

Feat/integration example

zihaolin96 committed:
sdk v0.1.4

zihaolin96 committed:
Merge pull request #87 from Klavis-AI/feat/0.1.4

sdk v0.1.4

zihaolin96 committed:
update

zihaolin96 committed:
Merge pull request #88 from Klavis-AI/feat/update_example_2

update example

zihaolin96 committed:
update

zihaolin96 committed:
Merge pull request #89 from Klavis-AI/lint/update

lint/update

zihaolin96 committed:
feat/use fern for sdk

zihaolin96 committed:
Merge pull request #90 from Klavis-AI/feat/update_sdk

feat/use fern for sdk

zihaolin96 committed:
a

zihaolin96 committed:
Merge pull request #91 from Klavis-AI/feat/update_fern

Add action for fern sdk generation

zihaolin96 committed:
a

zihaolin96 committed:
Merge pull request #92 from Klavis-AI/fix/rm_old_sdk_action

rm old sdk action

zihaolin96 committed:
update generator

zihaolin96 committed:
Merge pull request #93 from Klavis-AI/fix/update_generator

fern/update generator

zihaolin96 committed:
add get oauth url endpoint to sdk

zihaolin96 committed:
Merge pull request #94 from Klavis-AI/feat/upgrade_sdk

add get oauth url endpoint to sdk

zihaolin96 committed:
add release step in action

zihaolin96 committed:
Merge pull request #95 from Klavis-AI/feat/add_release_auto

add release step in action

zihaolin96 committed:
override fern config

zihaolin96 committed:
Merge pull request #96 from Klavis-AI/fix/override_fern_config

override fern config for api key

zihaolin96 committed:
update enum config for fern

zihaolin96 committed:
Merge pull request #97 from Klavis-AI/fix/config_enum_update

update enum config for fern

zihaolin96 committed:
override api key for fern sdk

zihaolin96 committed:
Merge pull request #98 from Klavis-AI/fix/fern_override_v2

override api key for fern sdk

zihaolin96 committed:
add openai example

zihaolin96 committed:
Merge pull request #99 from Klavis-AI/feat/add_openai_example

add openai example

zihaolin96 committed:
customize sdk name

zihaolin96 committed:
add metadata

zihaolin96 committed:
Merge pull request #100 from Klavis-AI/feat/fern_customize_method_name

customize sdk name

zihaolin96 committed:
update example

zihaolin96 committed:
add summary

zihaolin96 committed:
Merge pull request #101 from Klavis-AI/update/openai_example

update openai example

zihaolin96 committed:
update openai example

zihaolin96 committed:
Merge pull request #102 from Klavis-AI/update/openai_example_v2

update openai example

zihaolin96 committed:
add other models

zihaolin96 committed:
add ts sdk workflow

zihaolin96 committed:
Merge pull request #104 from Klavis-AI/feat/add_ts_sdk_workflow

add ts sdk workflow

zihaolin96 committed:
add ts sdk example with openai

zihaolin96 committed:
Merge pull request #105 from Klavis-AI/feat/use_ts_sdk_in_example

add ts sdk example with openai

zihaolin96 committed:
update readme with SDK

zihaolin96 committed:
rm files

zihaolin96 committed:
Merge pull request #107 from Klavis-AI/update/readme_with_sdk

Update/readme with sdk

zihaolin96 committed:
update readme

zihaolin96 committed:
Merge pull request #108 from Klavis-AI/update/readme_with_sdk_2

update readme

zihaolin96 committed:
update with new sdk

zihaolin96 committed:
Merge pull request #110 from Klavis-AI/update/examples

update with new sdk

zihaolin96 committed:
update

zihaolin96 committed:
Merge pull request #103 from Klavis-AI/feat/add_gemini_claude_example

add other models

zihaolin96 committed:
rm unrelated files

zihaolin96 committed:
Merge pull request #111 from Klavis-AI/fix/rm_unrelated_files

rm unrelated files

zihaolin96 committed:
add crewai

zihaolin96 committed:
Merge pull request #112 from Klavis-AI/feat/add_crew_ai

add crewai

xiangkaiz committed:
update readme.

xiangkaiz committed:
Update wordpress readme.

xiangkaiz committed:
Fix package.

xiangkaiz committed:
Fix lock.

zihaolin96 committed:
feat/add affinity

xiangkaiz committed:
Fix jira dockerfile.

xiangkaiz committed:
Fix notion dockerfile.

xiangkaiz committed:
Hotfix.

xiangkaiz committed:
Fix more dockerfiles.

xiangkaiz committed:
More fixes to build failures.

xiangkaiz committed:
Fix postgres package-lock.json.

xiangkaiz committed:
Fix supabase mcp startup.

zihaolin96 committed:
add python code for crew ai

zihaolin96 committed:
Merge pull request #114 from Klavis-AI/feat/add_crew_ai_python_example

add python code for crew ai

zihaolin96 committed:
optimize salesforce

zihaolin96 committed:
upgrade SDK

zihaolin96 committed:
Merge pull request #115 from Klavis-AI/feat/upgrade_sdk_0625

upgrade SDK

zihaolin96 committed:
update python

zihaolin96 committed:
add history to test client

zihaolin96 committed:
Merge pull request #116 from Klavis-AI/feat/salesforce_optimize

Feat/salesforce optimize

zihaolin96 committed:
quick fix on dockerfile

zihaolin96 committed:
Merge pull request #117 from Klavis-AI/fix/salesforce_update

quick fix on dockerfile

zihaolin96 committed:
salesforce tool optimize

zihaolin96 committed:
Merge pull request #118 from Klavis-AI/update/salesforce_tool_optimize

salesforce tool optimize

zihaolin96 committed:
update salesforce tools

zihaolin96 committed:
Merge pull request #119 from Klavis-AI/update/salesforce_tool_optimize_2

rm salesforce tools

xiangkaiz committed:
Add whatsapp MCP server.

xiangkaiz committed:
Fix dockerfile.

zihaolin96 committed:
rm default streamable_http

zihaolin96 committed:
Merge pull request #120 from Klavis-AI/rm_streamable_http_default

rm default streamable_http

zihaolin96 committed:
optimize salesforce tool

zihaolin96 committed:
Merge pull request #121 from Klavis-AI/feat/optimize_salesforce_tool

optimize salesforce tool

zihaolin96 committed:
add crewai demo

zihaolin96 committed:
Merge pull request #122 from Klavis-AI/feat/crewai_demo

feat: add crewai demo

zihaolin96 committed:
update tool

zihaolin96 committed:
Merge pull request #113 from Klavis-AI/feat/affinity

feat: add affinity

xiangkaiz committed:
Add roadmap section.

zihaolin96 committed:
feat: change affinity to v2

zihaolin96 committed:
Merge pull request #123 from Klavis-AI/update/optimize_affinity_tool

feat: change affinity to v2

dependabot[bot] committed:
Bump github.com/go-viper/mapstructure/v2 in /mcp_servers/github

Bumps [github.com/go-viper/mapstructure/v2](https://github.com/go-viper/mapstructure) from 2.2.1 to 2.3.0.
- [Release notes](https://github.com/go-viper/mapstructure/releases)
- [Changelog](https://github.com/go-viper/mapstructure/blob/main/CHANGELOG.md)
- [Commits](https://github.com/go-viper/mapstructure/compare/v2.2.1...v2.3.0)

---
updated-dependencies:
- dependency-name: github.com/go-viper/mapstructure/v2
  dependency-version: 2.3.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

zihaolin96 committed:
migrate docs

zihaolin96 committed:
Merge pull request #124 from Klavis-AI/dependabot/go_modules/mcp_servers/github/github.com/go-viper/mapstructure/v2-2.3.0

Bump github.com/go-viper/mapstructure/v2 from 2.2.1 to 2.3.0 in /mcp_servers/github

zihaolin96 committed:
Merge pull request #125 from Klavis-AI/feat/migrate_docs

migrate docs

zihaolin96 committed:
fix supabase and zendesk icon

zihaolin96 committed:
Merge pull request #126 from Klavis-AI/fix/fix_cards

fix supabase and zendesk icon

lokeswaran-aj committed:
feat(airtable): Add MCP server for Airtable

zihaolin96 committed:
rm lint workflow

zihaolin96 committed:
Merge pull request #128 from Klavis-AI/fix/rm_gh_workflow

rm lint workflow

zihaolin96 committed:
add affinity doc

lokeswaran-aj committed:
feat(airtable): Add Dockerfile for MCP Airtable server

lokeswaran-aj committed:
refactor(airtable): Move the tools and make request code to tools folder

lokeswaran-aj committed:
feat(airtable): Add tool to retrieve table information from Airtable

lokeswaran-aj committed:
feat(airtable): Add create_table tool for creating new tables in Airtable

lokeswaran-aj committed:
feat(airtable): Add update_table tool for modifying existing tables in Airtable

lokeswaran-aj committed:
feat(airtable): Add list_records tool for retrieving all records from a specified table

lokeswaran-aj committed:
feat(airtable): Add get_record tool for retrieving a single record from a specified table

lokeswaran-aj committed:
feat(airtable): Add create_records tool for creating multiple records in a specified table

lokeswaran-aj committed:
feat(airtable): Add delete_records tool for removing multiple records from a specified table

lokeswaran-aj committed:
fix(airtable): Enhance error handling in Airtable tools and update logging configuration

lokeswaran-aj committed:
feat(airtable): Add create_field tool for creating new fields in a specified table

lokeswaran-aj committed:
feat(airtable): Add update_records tool for updating multiple records with optional upsert functionality

lokeswaran-aj committed:
feat(airtable): Add update_field tool for modifying existing fields in a specified table

lokeswaran-aj committed:
feat(airtable): Enhance list_records tool with optional parameters for filtering, sorting, and field selection

lokeswaran-aj committed:
refactor(airtable): Move the argument description to the tool definition

lokeswaran-aj committed:
refactor(airtable): Simplify Dockerfile by adjusting file paths for better organization

lokeswaran-aj committed:
feat(airtable): Add README.md for Airtable MCP server

lokeswaran-aj committed:
refactor(airtable): Use the argument's get method to get the value

zihaolin96 committed:
Merge pull request #129 from Klavis-AI/feat/add_affinity_doc

add affinity doc and fix logo

zihaolin96 committed:
add search for affinity

zihaolin96 committed:
Merge pull request #130 from Klavis-AI/feat/add_affinity_api

add search for affinity

zihaolin96 committed:
add search for affinity

zihaolin96 committed:
Merge pull request #131 from Klavis-AI/update/affinity_tool

docs: add search for affinity

zihaolin96 committed:
update auth logic to support both dev and prod

zihaolin96 committed:
Merge pull request #127 from lokeswaran-aj/feat/airtable

feat(airtable): Add MCP server for Airtable

zihaolin96 committed:
fix airtable docke

zihaolin96 committed:
Merge pull request #132 from Klavis-AI/fix/airtable_dockerfile

fix airtable docke

zihaolin96 committed:
update readme

zihaolin96 committed:
update readme (#133)

zihaolin96 committed:
doc: airtable

zihaolin96 committed:
doc: airtable (#134)

zihaolin96 committed:
fix logo in docs

zihaolin96 committed:
docs: fix logo (#135)

zihaolin96 committed:
add cards category

zihaolin96 committed:
docs: add cards category (#137)

Mayank-MSJ-Singh committed:
Adding Hubspot CRM MCP (#109)

Co-authored-by: Zihao Lin <zihaolin@klavis.ai>

zihaolin96 committed:
fix: hubspot (#138)

dependabot[bot] committed:
build(deps): bump urllib3 from 2.4.0 to 2.5.0 in /mcp_servers/hubspot (#139)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zihaolin96 committed:
docs: hubspot (#140)

zihaolin96 committed:
docs: Add no code part (#141)

zihaolin96 committed:
feat: add affinity file tools (#142)

zihaolin96 committed:
Revert "feat: add affinity file tools" (#143)

lokeswaran-aj committed:
feat(monday): Add Monday MCP server (#136)

lokeswaran-aj committed:
docs(monday): Add OAuth endpoints and add Monday MCP server documentation (#144)

zihaolin96 committed:
feat: add get auth metadata api to docs (#146)

zihaolin96 committed:
feat: update google genai examples (#148)

askdevai-bot committed:
docs: add LLM-friendly API documentation (#149)

zihaolin96 committed:
update: hubspot - add create note tool (#150)

zihaolin96 committed:
docs: update white label with SDK (#152)

zihaolin96 committed:
feat: add mastra example (#153)

zihaolin96 committed:
update: rm connection type (#155)

zihaolin96 committed:
docs: add list tool docs (#156)

zihaolin96 committed:
fix: gmail tool name (#157)

zihaolin96 committed:
examples: add together ai demo (#158)

zihaolin96 committed:
add kiro docs (#159)

zihaolin96 committed:
update: add gcalendar new tool - invite attendees (#160)

zihaolin96 committed:
update: readme (#162)

zihaolin96 committed:
upgrade mcp denpendency (#163)

zihaolin96 committed:
update: gcalendar attendee (#164)

zihaolin96 committed:
update: support case-insensitive for server name (#166)

zihaolin96 committed:
rm workflow (#167)

zihaolin96 committed:
doc: add crewai and llamaindex post (#168)

zihaolin96 committed:
doc: add mastra (#169)

udaykakade25 committed:
feat(linkedin): Add MCP server for LinkedIn (#165)

zihaolin96 committed:
fix: try catch get notes error (#171)

dependabot[bot] committed:
build(deps): bump form-data from 4.0.2 to 4.0.4 in /mcp_servers/notion (#172)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump form-data from 4.0.2 to 4.0.4 in /mcp_servers/firecrawl_deep_research (#173)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump starlette from 0.46.2 to 0.47.2 in /mcp_servers/hubspot (#174)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump axios from 1.10.0 to 1.11.0 in /examples/mastra (#175)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zihaolin96 committed:
fix: Affinity get notes json error (#176)

zihaolin96 committed:
doc: readme update (#177)

zihaolin96 committed:
fix: linkedin prod issue (#178)

dependabot[bot] committed:
build(deps): bump form-data from 4.0.2 to 4.0.4 in /mcp_servers/firecrawl (#179)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

udaykakade25 committed:
docs(linkedin): Add LinkedIn Documentation (#170)

zihaolin96 committed:
fix: update linkedin dockerfile (#182)

dependabot[bot] committed:
build(deps): bump form-data from 4.0.3 to 4.0.4 in /examples/openai/typescript (#181)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump form-data from 4.0.3 to 4.0.4 in /examples/google-genai/typescript (#183)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zihaolin96 committed:
doc: add linkedin oauth doc (#180)

dependabot[bot] committed:
build(deps): bump starlette from 0.40.0 to 0.47.2 in /mcp_servers/google_slides (#184)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

zihaolin96 committed:
fix: linkedin tool issue (#185)

udaykakade25 committed:
refactor(linkedin): Removes `get_company_info` and `get_user_posts` (#186)

zihaolin96 committed:
doc: update linkedin tool (#189)

zihaolin96 committed:
Revert "doc: update linkedin tool" (#190)

udaykakade25 committed:
docs(linkedin): Removed `get_company_info` and `get_user_posts` from docs (#188)

zihaolin96 committed:
doc: update doc (#191)

zihaolin96 committed:
update: gcalendar create event support send notification (#192)

zihaolin96 committed:
update: support google meet for gcalendar (#193)

xiangkaiz committed:
Add build guide.

zihaolin96 committed:
feat: revamp notion (#195)

zihaolin96 committed:
fix: update dockerfile for notion (#196)

xz-dev committed:
feat(Dropbox): Init Dropbox MCP server (#145)

xiangkaiz committed:
fix docs.

zihaolin96 committed:
update: readme roadmap (#201)

zihaolin96 committed:
doc: add canva oauth doc (#202)

zihaolin96 committed:
doc: Add develop with AI (#208)

zihaolin96 committed:
doc: add image (#209)

zihaolin96 committed:
doc: add xero oauth doc (#210)

zihaolin96 committed:
doc: xero doc update (#211)

zihaolin96 committed:
update openapi json (#212)

xz-dev committed:
docs/MCP_SERVER_GUIDE: Additional Resources and Documentation Guidelines (#203)

udaykakade25 committed:
feat(linkedin): Adds 2 new tool in Linkedin MCP Server (#194)

zihaolin96 committed:
doc: Introduce Klavis Security (#214)

xiangkaiz committed:
Update doc. (#215)

xiangkaiz committed:
Doc (#216)

Mayank-MSJ-Singh committed:
brave-search addition (#206)

zihaolin96 committed:
doc: add brave search doc (#218)

zihaolin96 committed:
doc: update logo (#219)

zihaolin96 committed:
update: upgrade openapi for sdk (#220)

zihaolin96 committed:
doc: add dropbox doc (#222)

xiangkaiz committed:
doc: Add unified MCP server endpoint and fix openapi.json (#226)

xiangkaiz committed:
Doc: Add tutorial to use the unified mcp server. (#227)

xiangkaiz committed:
Doc: Replace with mermaid code. (#228)

xiangkaiz committed:
Doc: Fix unified MCP server docs. (#229)

zihaolin96 committed:
doc: add onboarding doc (#230)

raghavendra2222 committed:
feat: Added new MCP server for Twilio integration

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Klavis-AI/klavis?pullRequest=231) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Klavis-AI/klavis?pullRequest=231) before we can accept your contribution.<br/>**6** out of **7** committers have signed the CLA.<br/><br/>:white_check_mark: Mayank-MSJ-Singh<br/>:white_check_mark: xiangkaiz<br/>:white_check_mark: xz-dev<br/>:white_check_mark: raghavendra2222<br/>:white_check_mark: udaykakade25<br/>:white_check_mark: zihaolin96<br/>:x: dependabot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Klavis-AI/klavis?pullRequest=231) it.</sub>

raghavendra2222 committed:
feat: Add new tools and align server with project standards

raghavendra2222 reviewed (commented):
<img width="1027" height="484" alt="get_message_details" src="https://github.com/user-attachments/assets/5f1ed0d2-9c13-47d3-ad13-d0a677ecbf88" />
<img width="1028" height="770" alt="list_messages" src="https://github.com/user-attachments/assets/8292d03d-7358-4014-841b-55af48d547e5" />
<img width="1027" height="768" alt="list_phone_numbers" src="https://github.com/user-attachments/assets/0f51695d-e152-4fd4-860a-21578fe5521e" />
<img width="1021" height="426" alt="redact_message" src="https://github.com/user-attachments/assets/06ebe1f0-d185-4e0e-b8be-e67b917505fc" />
<img width="1027" height="520" alt="send_MMS" src="https://github.com/user-attachments/assets/253e9987-b6ba-48fc-ad67-2788222ed79d" />
<img width="1028" height="514" alt="send_SMS" src="https://github.com/user-attachments/assets/63ddb3f0-c1df-45b3-af83-1c2a4a982242" />

Thank you for the detailed feedback. I have completed all the requested changes and added several new features as suggested.

- The project has been restructured to match the `affinity` server example.
- Error handling, authentication, and other best practices have been implemented.
- I have added three new tools: `send_mms`, `redact_message`, and `list_phone_numbers`.

Please see the new "Proof of Correctness" for all the new tools below.

**Tool: `list_phone_numbers`**
*(Drag and drop your list_phone_numbers screenshot here)*

**Tool: `send_mms`**
*(Drag and drop your send_mms screenshot here)*

**Tool: `redact_message` (with verification)**
*(Drag and drop your redact_message screenshot here)*

I've also updated the `README.md` to reflect all these changes. Please let me know if there's anything else required.

raghavendra2222 commented:
Hi Uday,

I’ve made the requested changes to the PR and submitted it for review.

However, I noticed that the status of the messages is showing as
*undelivered*, and the platform indicates the following:

“You have Business Registration ID (such as a US EIN)”
“You are an individual and not acting as part of a registered business or
organization”

As I’m an individual contributor and not operating under a registered
business, this seems to be causing the issue.

On Sun, Aug 3, 2025 at 9:30 AM Uday Kakade ***@***.***> wrote:

> *udaykakade25* left a comment (Klavis-AI/klavis#231)
> <https://github.com/Klavis-AI/klavis/pull/231#issuecomment-3148535401>
>
> Hi @raghavendra2222 <https://github.com/raghavendra2222>, I reviewed your
> PR and below are changes you need to make.
>
>    1. For MCP Server, folder structure should align with Klavis AI's
>    style and format (reference:
>    https://github.com/Klavis-AI/klavis/tree/main/mcp_servers/affinity)
>    2. Also, attach screenshots (or a demo video) of the tools in the
>    comment.
>    3. Try to add multiple tools (here, you listed just one tool - send_sms
>    )
>
> Please read the contributing guidelines and sign Contributor License
> Agreement before making a PR.
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Klavis-AI/klavis/pull/231#issuecomment-3148535401>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJWS3TYBMGV336HFNB3DRED3LY2JRAVCNFSM6AAAAACC7KEB5OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNBYGUZTKNBQGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

raghavendra2222 committed:
feat: Refactor Twilio MCP server to use proper MCP infrastructure

- Restructured to follow Klavis AI folder structure (mcp_servers/twilio/)
- Implemented proper MCP server using mcp.server.lowlevel.Server
- Added comprehensive tool set: send_sms, send_mms, list_messages, get_message_details, redact_message, list_phone_numbers
- Fixed HTTP transport and authentication middleware
- Added proper error handling and logging
- Updated Dockerfile and requirements.txt
- Added comprehensive README.md and test_sms.py
- Resolves PR feedback about MCP infrastructure and folder structure

raghavendra2222 reviewed (commented):
<img width="992" height="806" alt="get_message_details" src="https://github.com/user-attachments/assets/3d61e5bc-1fee-4494-83c7-ffb707605329" />
<img width="997" height="804" alt="list_messages" src="https://github.com/user-attachments/assets/93cf9307-aa47-496d-b5db-acb518778b08" />
<img width="996" height="799" alt="list_phone_numbers" src="https://github.com/user-attachments/assets/f7b578b4-7b53-4dea-8746-21adf396b059" />
<img width="986" height="805" alt="redact_message" src="https://github.com/user-attachments/assets/a536446d-e5b3-41df-9f3d-00586aaa0390" />
<img width="1001" height="806" alt="send_mms" src="https://github.com/user-attachments/assets/0886ce4e-293d-4271-8b7b-d7ac7a3ceaba" />
<img width="996" height="802" alt="send_sms" src="https://github.com/user-attachments/assets/b707c3d0-3274-4e9e-b7ff-3aaf6a932f8c" />

I have done a complete refactor of my submission based on your recommendation and a deep dive into the affinity server's architecture. I've discarded the previous standalone approach and have now implemented the Twilio integration as a proper, modular Python package, ready to be consumed by the MCP orchestrator.
Here is a summary of the new architecture:
-> Centralized Logic: All core Twilio API interaction is now encapsulated within a TwilioClient class in base.py. This class handles authentication and all API operations.
-> Modular Tools: Each tool (send_sms, list_messages, etc.) is now a clean, separate async function located in feature-specific files (messages.py, phone_numbers.py). These act as thin wrappers around the central client.
-> Clear Entry Point: The root __init__.py now exposes all the tool functions as the public API of this package, which is what the MCP infrastructure will discover and use.
-> Flexible Authentication: The client can now be initialized using either environment variables or a request-specific auth token passed via context, making it much more robust.
I believe this new structure fully aligns with the project's standards and correctly uses the MCP framework. The code is ready for re-review when you have a moment.

zihaolin96 commented on mcp_servers/twilio/Dockerfile:
```suggestion
EXPOSE 5000
```

zihaolin96 commented on mcp_servers/twilio/server.py:
nit: could you help follow the same format as other files like this - https://github.com/Klavis-AI/klavis/blob/main/mcp_servers/affinity/server.py#L1300

zihaolin96 commented on mcp_servers/twilio/test_sms.py:
let's remove this file or create a new folder `test` for this

raghavendra2222 committed:
Refactor Twilio MCP server to match affinity server format

- Change default port from 5001 to 5000 across all files
- Update server.py to follow exact affinity server structure:
  * Add proper SSE and StreamableHTTP transports
  * Implement try-catch error handling for all tools
  * Use standard MCP routes (/sse, /messages/, /mcp)
  * Remove custom /http endpoint
- Update README.md to reflect standard MCP endpoints
- Remove test_sms.py (used non-standard HTTP endpoints)
- Ensure consistent code formatting with affinity server

raghavendra2222 reviewed (commented):
Refactor: Align server with Affinity MCP standard per review

Based on the recent feedback, this commit refactors the server to align it with the standard Affinity pattern and resolves several outstanding issues.

*   **Standardize Server Structure:** The application initialization, transports (SSE, StreamableHTTP), and lifespan management now directly follow the `mcp_servers/affinity/server.py` implementation for better consistency and reliability.

*   **Improve Error Handling:** All tool calls are now wrapped in `try...except` blocks to prevent crashes and provide better diagnostic logging with `logger.exception()`.

*   **Simplify and Clean Up:** Removed the non-standard `/http` endpoint and its associated test file (`test_sms.py`) to simplify the codebase. The README has been updated to only document the standard MCP endpoints.

*   **Bug Fixes:** Resolved a `SyntaxError` in the async SSE generator, added missing imports, and configured the application to use port 5000.

raghavendra2222 commented:
Hi Zihao,

I have made the changes and submitted for review

https://github.com/Klavis-AI/klavis/pull/231/commits/c738cb53a5831c4e56f08b8404c42d3d4b6c4837

On Fri, Aug 8, 2025 at 9:52 AM Zihao Lin ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In mcp_servers/twilio/test_sms.py
> <https://github.com/Klavis-AI/klavis/pull/231#discussion_r2263551063>:
>
> > @@ -0,0 +1,123 @@
> +#!/usr/bin/env python3
>
> let's remove this file or create a new folder test for this
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Klavis-AI/klavis/pull/231#pullrequestreview-3101576087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJWS3T3BNYBVVDPS3AGBRBT3MTIUDAVCNFSM6AAAAACC7KEB5OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMBRGU3TMMBYG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
